sst/test/src/tg2_cache_access.cc

```diff
@@ -0,0 +1,301 @@
+  uint64_t firstLoadData = reads[1]->data_;
+  // check the last entry of the vector to avoid load executed by branch
+  // prediction.
+  uint64_t lastLoadCycleLatency =
+      reads[reads.size() - 1]->endCycle_ - reads[reads.size() - 1]->startCycle_;
+  uint64_t lastLoadData = reads[reads.size() - 1]->data_;
+
+  EXPECT_LT(lastLoadCycleLatency, firstLoadCycleLatency);
+  EXPECT_EQ(firstLoadData, lastLoadData);
+}
+
+TEST_CASE(TG2, "load_after_store_on_same_address_should_be_return_from_cache",
+          "src",
+          R"(
+    # Get heap address
+    mov x0, 0
+    mov x8, 214
+    svc #0
+
+    # Setup
+    mov x2, #2048
+
+    # Load from heap - done to determine max clock cycle without any caching.
+    ldr x1, [x0]
+
+    # Store value in x3 at address in x2
+    str x1, [x2]
+
+    # Multiple instructions to wait for store to execute.
+    # Can't be replaced by loop to ensure that caching was
+    # established by store
+    mov x4, #0
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+    add x4, x4, #1
+
+    # last load
+    ldr x4, [x2]
+
+    )",
+          "1024") {
+  Parser p = Parser(capturedStdout);
+  std::vector<ParsedMemRead*> reads = p.getParsedMemReads();
+  // skip first parsed request as that one will be caused by heap address
+  // retrieval into x0.
+  uint64_t firstLoadCycleLatency = reads[1]->endCycle_ - reads[1]->startCycle_;
+  uint64_t firstLoadData = reads[1]->data_;
+  uint64_t lastLoadCycleLatency = reads[2]->endCycle_ - reads[2]->startCycle_;
+  uint64_t lastLoadData = reads[2]->data_;
+
+  EXPECT_LT(lastLoadCycleLatency, firstLoadCycleLatency);
+  EXPECT_EQ(1024, firstLoadData);
+  EXPECT_EQ(1024, lastLoadData);
+}
+
+TEST_CASE(TG2,
+          "multiple_loads_after_stores_on_same_address_should_return_"
+          "from_cache",
+          "src",
+          R"(
+    # Get heap address
+    mov x0, 0
+    mov x8, 214
+    svc #0
+
+    # Setup
+    mov x2, #2048
+
+    # Load from heap - done to determine max clock cycle without any caching.
+    ldr x1, [x0]
+
+    # Store value in x3 at address in x2
+    str x1, [x2]
+    str x1, [x2, #8]
+    str x1, [x2, #16]
+    str x1, [x2, #24]
+    str x1, [x2, #32]
+
+    # Multiple instructions to wait for all stores to execute.
+    # Can't be replaced by loop to ensure that caching was
+    # established by store. This sequence is larger to ensure all
+    # store instructions have been executed before we execute load
+    # instructions.
+    mov x4, #0
```

Comment:
I wonder if you could use a loop but force an exception to be invoked (i.e. syscall) just before the branch call to ensure a pipeline flush? If it isn't dependant on anything then it should be executed near the start of every loop

Reply:
Even if a flush is caused before the branch, the problem lies more with the branch prediction mechanism, an incorrect prediction will lead to the execution of the last load instruction prematurely i.e before the first request completes. Since we can't flush SST that request will still propagate and the result will not be used, however, it will still be cached in SST.

Reply:
My thinking was that the exception caused will essentially prevent the prediction from going through as the syscall will stall and flush the pipeline until it has finished executing - if the loop is constructed correctly. Then by the time the load instruction is fetched, the branch should have executed and the actual path would be known.

Something as simple as the following may work:
```
mov x0, xzr
mov x1, #100
add x0, x1
cmp x0, x1
mrs x10, CNTVCT_el0
b.lt -0x12

ldr x1, [x0]
```

Additionally, you can take advantage of the default branch behaviour - if taken is always the first guess then the above code will work regardless. If it is not-taken then you can get creative with jumping over the final load then jumping back to it based on a taken condition.

May be worth a small trial to at least see the behaviour it exhibits? Would be good if there was a way that cut down on this "padding code".

Reply:
Please check